{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Finding Genes\n",
    "\n",
    "Now that we have run RepeatMasker on our assembled genome, we are going to identify coding genes. To do this, we can\n",
    "generally take a few different approaches:\n",
    "\n",
    "1. Using RNA-seq data that we generated for our organism, which we can align to the genome, and then use it to identify\n",
    "genes.\n",
    "\n",
    "2. Using another similar organism for which we already have information on what genes should \"look like\". For example,\n",
    "we can use already identified human genes to identify genes in a chimpanzee genome.\n",
    "\n",
    "Today, we are going to use the first approach. In this section we are going to use RNA-seq data to identify genes in four basic steps:\n",
    "\n",
    "1. Align RNA-seq data to our assembly.\n",
    "2. Identify putative introns in these alignments.\n",
    "3. Train a model that tells Augustus what a gene in our assembly looks like.\n",
    "4. Identify genes with Augustus."
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Aligning RNA-seq data\n",
    "\n",
    "For our input RNA-seq data we are going to use sequencing generated by a _P. falciparum_ project performed in part at the\n",
    "Wellcome Sanger Institute in Cambridge, UK. We are going to use this data to help us build a model of what a gene in our\n",
    " assembled genome should look like. This data comes from the following paper:\n",
    "\n",
    "Lia Chappell, Philipp Ross, Lindsey Orchard, Timothy J. Russell, Thomas D. Otto, Matthew Berriman, Julian C. Rayner, and Manuel LlinÃ¡s. _**Refining the transcriptome of the human malaria parasite Plasmodium falciparum using amplification-free RNA-seq**_. BMC Genomics, 2020.\n",
    "\n",
    "To generate the data used by this module, we accessed the sequencing data generated by the above paper on the\n",
    "[European Nucleotide Archive](https://www.ebi.ac.uk/ena/browser/home) under the accession `ERX2287076`. This accession represents\n",
    "paired-end Illumina-based short-read sequencing data generated for a single _P. falciparum_ isolate. We have already\n",
    "filtered the reads to those which should align to the assembly that you generated as part of the assembly module.\n",
    "\n",
    "Take a look at the data in your terminal:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "zcat Pfalc_chr5.1.fq.gz | head -n 12"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command should show you the first 3 reads in this file.\n",
    "\n",
    "You should have already learned how to align RNA-seq data as part of the RNA-seq module. We are going to use the same\n",
    "approach but for our assembly.\n",
    "\n",
    "First, we have to index our assembly we made yesterday. This command makes it suitable for RNA-seq alignment with `hisat2`:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "hisat2-build PB.masked.fasta PB.masked.fasta.idx"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now we are going to align our reads to our reference genome:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "hisat2 --max-intronlen 10000 -x PB.masked.fasta.idx -1 Pfalc_chr5.1.fq.gz -2 Pfalc_chr5.2.fq.gz > Pfalc.sam"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now, convert the alignment to bam:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "samtools view -Sbo Pfalc.bam Pfalc.sam"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "Our gene finding approach that we will perform later requires an extra processing step. This tool will filter out\n",
    "alignments that align to our assembly more than once:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "filterBam --uniq --in Pfalc.bam --out Pfalc.ssf.bam --paired --pairwiseAlignment"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "The `--paired` and `--pairwiseAlignment` tags tell `filterBam` that we are using paired-end data.\n",
    "\n",
    "Now we sort and index the file as we did in the RNA-seq module:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "samtools sort -o Pfalc.sorted.ssf.bam Pfalc.ssf.bam\n",
    "\n",
    "samtools index Pfalc.sorted.ssf.bam"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "We will take a look at these reads in IGV later to see how well our alignment worked.\n",
    "\n",
    "We are now ready to start gene discovery."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Converting RNA-seq data to intron information\n",
    "\n",
    "We are now going to start using the Augustus tool. Augustus is available on github:\n",
    "\n",
    "https://github.com/Gaius-Augustus/Augustus\n",
    "\n",
    "Augustus was originally published as a free web-tool in the following paper:\n",
    "\n",
    "  Mario Stanke and Burkhard Morgenstern. _**AUGUSTUS: a web server for gene prediction in eukaryotes that allows user-defined constraints**_. Nucleic Acids Research (2005).\n",
    "\n",
    "To find genes, Augustus needs information on what a gene \"looks like\" in the organism that you have assembled. In\n",
    "Eukaryotic genomes (such as _P. falciparum_), the most complicated part of gene discovery typically is identifying\n",
    "introns/splice sites. Augustus thus uses the RNA-seq data we generated above to identify patterns in the DNA sequence\n",
    "which are indicative of splicing.\n",
    "\n",
    "This practical follows several parts of a tutorial developed by the Augustus authors which was published as part of the\n",
    "following paper:\n",
    "\n",
    "Katharina J. Hoff and Mario Stanke. _**Predicting Genes in Single Genomes with AUGUSTUS**_. Current Protocols in Bioinformatics (2018).\n",
    "\n",
    "A free preprint of this publication, which is very similar to the published version, is available using\n",
    "[this link](https://math-inf.uni-greifswald.de/storages/uni-greifswald/fakultaet/mnf/mathinf/stanke/augustus_wrp.pdf).\n",
    "\n",
    "This workflow also makes use of several scripts from the tool BRAKER, also written by the Augustus authors. We have\n",
    "copied the scripts required for this tutorial from BRAKER's [github repository](https://github.com/Gaius-Augustus/BRAKER)\n",
    "for you to use today and placed them in the `~/course_data/annotation/scripts/` folder.\n",
    "\n",
    "**Please note:** The method presented here is a simplification of the approach that you would need to take to accurately\n",
    " predict genes in a new genome and is presented here as a foundation on which to build. We have omitted several steps\n",
    " for the sake of computing resources and time. See the above Augustus tutorial for further steps that can be taken to\n",
    " more accurately predict genes in genome assemblies. There are other tools and resources that can be used\n",
    " to annotate both eukaryotic and prokaryotic genomes.\n",
    "\n",
    "First, we need to find all the introns in our aligned RNA-sequencing data and then filter out low-quality introns:"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "bam2hints --intronsonly --in=Pfalc.sorted.ssf.bam --out=Pfalc.introns.gff\n",
    "\n",
    "filterIntronsFindStrand.pl PB.masked.fasta Pfalc.introns.gff --score > Pfalc.introns.f.gff"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Next, we would normally use the [\"GeneMark-ES\"](http://exon.gatech.edu/GeneMark/gmes_instructions.html) software to\n",
    "identify genes using our intron predictions. Unfortunately, the licence of GeneMark-ES prevents us from putting this\n",
    "software on a virtual machine. As such, we have pre-run \"GeneMark-ES\" on the data in this module and provided the\n",
    "necessary output for you already in the file `genemark.gtf`.\n",
    "\n",
    "To generate this file we used the following command (Do not run this command!):"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "gmes_petap.pl --verbose --sequence=PB.masked.fasta --ET=Pfalc.introns.f.gff --soft_mask 1000"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "While we cannot demonstrate how this code runs today, you can go to the\n",
    "[GeneMark-ES website](http://exon.gatech.edu/GeneMark/gmes_instructions.html) and download a copy for your own personal\n",
    "use after this course.\n",
    "\n",
    "Let's take a look at the output of GeneMark-ES and see what it contains:"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "source": [
    "head -n 31 genemark.gtf"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command should print predictions for the first two genes identified by GeneMark-ES, fittingly named \"1_g\" and \"2_g\".\n",
    "\n",
    "**Questions:**\n",
    "\n",
    "1. How many exons does the gene \"1_g\" have?\n",
    "2. Can you think of a simple LINUX command to figure out how many genes GeneMark-ES identified?\n",
    "3. How many genes did GeneMark find?\n",
    "\n",
    "Next, we are going to use various commands from [BRAKER](https://github.com/Gaius-Augustus/BRAKER) to generate a list of \"high quality\" genes using the list of introns (`Pfalc.introns.f.gff`) and gene predictions from GeneMark-ES (`genemark.gtf`) that you made above:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "filterGenemark.pl genemark.gtf Pfalc.introns.f.gff"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command will generate two files: `genemark.f.good.gtf` and `genemark.f.bad.gtf` that we will use later.\n",
    "\n",
    "This command should also give output like (actual output may differ slightly):\n",
    "\n",
    "    Number of cds hints is 0\n",
    "    Average gene length: 2614\n",
    "    Average number of introns: 1.89261744966443\n",
    "    Good gene rate: 0.422818791946309\n",
    "    Number of genes: 300\n",
    "    Number of complete genes: 298\n",
    "    Number of good genes: 126\n",
    "    Number of one-exon-genes: 104\n",
    "    Number of bad genes: 174\n",
    "    Good intron rate: 0.200570201421801\n",
    "    One exon gene rate (of good genes): 0.349206349206349\n",
    "    One exon gene rate (of all genes): 0.346666666666667\n",
    "\n",
    "This output is telling us that, out of the total number of genes predicted by GeneMark-ES (300), only 126 look like\n",
    "real genes according to Augustus.\n",
    "\n",
    "Next, we are going to use scripts provided as part of Augustus to further filter our predicted genes down to a set of reliable hits. Run the following command:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "computeFlankingRegion.pl genemark.f.good.gtf"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command is simply calculating the total length of genes from our predictions above. It should generate an output similar to:\n",
    "\n",
    "    Total length gene length (including introns): 220281. Number of genes: 321.\n",
    "    Average Length: 686.233644859813\n",
    "\n",
    "    The flanking_DNA value is: 343 (the Minimum of 10 000 and 343)\n",
    "\n",
    "This is telling us that our predicted genes are, on average, about 670 basepairs long. We now need to use the value\n",
    "provided by this script, or approximately half the mean gene length (343) in our following commands:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "gff2gbSmallDNA.pl genemark.gtf PB.masked.fasta 343 tmp.gb 2> /dev/null\n",
    "\n",
    "filterGenesIn_mRNAname.pl genemark.f.good.gtf tmp.gb > bonafide.gb"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This will generate our list of likely true genes (`bonafide.gb`) so that we can get to identifying actual genes in our\n",
    "assembly. This file is in \"genbank\" format, the same format used by NCBI's GenBank to store sequence data about genes.\n",
    "You can go to GenBank with the following link: https://www.ncbi.nlm.nih.gov/genbank/\n",
    "\n",
    "**Questions:**\n",
    "\n",
    "1. What is the part of the command `2> /dev/null` actually doing?\n",
    "2. How many genes are in our final set of possible genes (`bonafide.gb`)?\n",
    "\n",
    "We are now ready to train Augustus to actually find genes. If these commands did not work, we have placed a backup of\n",
    "`bonafide.gb` in the `annotation_backups/` folder."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Training Augustus for Gene Finding\n",
    "\n",
    "We are now going to use our set of likely genes to \"train\" augustus on what a gene in our _P. falciparum_ assembly\n",
    "looks like. This approach uses a test-training method, where we split our genes into two pieces and \"train\" augustus on\n",
    "one half, and \"test\" on the other.\n",
    "\n",
    "This approach allows us to remove statistical circularity. In other words, we can see how good our predictor is since\n",
    "we don't use our testing set to train the data.\n",
    "\n",
    "We have already created some files in the `species/` directory for you to use for this exercise. But we first need to\n",
    "tell augustus that this directory is where we want our training data to go:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "export AUGUSTUS_CONFIG_PATH=/home/manager/course_data/annotation/data/"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This is known in linux as setting an \"environment variable\". This allows programs (like Augustus) to know certain\n",
    "pieces of information when they run. You can see that the environment variable is set properly by typing:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "echo $AUGUSTUS_CONFIG_PATH"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This should print the same path that we set above.\n",
    "\n",
    "We can now tell Augustus that we are going to train a new model, and Augustus will automatically place default files\n",
    "in the directory `species/mygenome/`:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "new_species.pl --species=mygenome\n",
    "\n",
    "etraining --species=mygenome bonafide.gb"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "You can put any name for `--species`, but for today we will use \"mygenome\". The first command sets up a working directory, and the second does an initial round of training on our genes. After running this command run the following:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "ls species/mygenome/"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "You should see the following output:\n",
    "\n",
    "    mygenome_exon_probs.pbl    mygenome_intron_probs.pbl  mygenome_metapars.cgp.cfg\n",
    "    mygenome_parameters.cfg    mygenome_igenic_probs.pbl  mygenome_metapars.cfg\n",
    "    mygenome_metapars.utr.cfg  mygenome_weightmatrix.txt\n",
    "\n",
    "These files contain information Augustus needs to identify genes. We are now going to identify better parameters using our test/training approach. First, we need to make a test and training set like we described above. In this case, we are going to simply split our data in half. But how many genes is that? To find out how many genes we have do:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "grep -c LOCUS bonafide.gb"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This should give \"126\" (if it doesn't don't worry, just change the following accordingly). So, if we are going to split\n",
    "50/50 to form train/test sets, we need to put 63 genes in both our training and test set (126 / 2 = 63). The\n",
    "following command will do this for us:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "randomSplit.pl bonafide.gb 63"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This will create two files: `bonafide.gb.train` and `bonafide.gb.test`.\n",
    "\n",
    "Now, lets run training on our training set:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "etraining --species=mygenome bonafide.gb.train &> etrain.out"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "What new information did this teach us? Most importantly, it gave us information on how often different stop\n",
    "codons are used. We can get this information from the \"tail\" of the output file with the following command:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "tail -n 6 etrain.out"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command simply prints the last 6 lines of this file, and you should see something like:\n",
    "\n",
    "**Note:** Your numbers for the below _will be different_. This is because the output of the `randomSplit.pl` command is...\n",
    "random!\n",
    "\n",
    "    tag:    8 (0.127)\n",
    "    taa:   44 (0.698)\n",
    "    tga:   11 (0.175)\n",
    "    end *EXON*\n",
    "    Storing parameters to file...\n",
    "    Writing exon model parameters [1] to file\n",
    "    /home/manager/course_data/annotation/data/species/mygenome/mygenome_exon_probs.pbl.\n",
    "\n",
    "This command takes care of most of the training, but we need to manually adjust a few parameters on our own.\n",
    "We are going to use a text editor known as `gedit` to change the training parameters according to the above data.\n",
    "Open the file we need to change:"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "source": [
    "gedit species/mygenome/mygenome_parameters.cfg"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "Now find the lines by scrolling through the file our using \"find\" (ctrl+f) that look like:\n",
    "\n",
    "    /Constant/amberprob                   0.33   # Prob(stop codon = tag), if 0 tag is assumed to ...\n",
    "    /Constant/ochreprob                   0.33   # Prob(stop codon = taa), if 0 taa is assumed to ...\n",
    "    /Constant/opalprob                    0.34   # Prob(stop codon = tga), if 0 tga is assumed to ...\n",
    "\n",
    "And change them to match the output of our command above, like:\n",
    "\n",
    "    /Constant/amberprob                   0.127   # Prob(stop codon = tag), if 0 tag is assumed to ...\n",
    "    /Constant/ochreprob                   0.698   # Prob(stop codon = taa), if 0 taa is assumed to ...\n",
    "    /Constant/opalprob                    0.175   # Prob(stop codon = tga), if 0 tga is assumed to ...\n",
    "\n",
    "Make sure you *save* the file before exiting - by using either the \"Save\" button at the top-right or by hitting \"ctrl+s\" on your keyboard.\n",
    "\n",
    "Finally, we are going to run Augustus on our test set and see how good our prediction is!"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "augustus --species=mygenome bonafide.gb.test > test.out"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "So, how did we do? The last 41 lines of the file `test.out` tell us exactly how well we did:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "tail -n 41 test.out"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This file provides 4 different tables, 1 for each of:\n",
    "\n",
    "1. Nucleotide-level prediction - are individual bases predicted to be part of genes, found within genes?\n",
    "2. Exon-level prediction - are individual exons predicted properly?\n",
    "3. Transcript-level prediction - are entire transcripts predicted properly?\n",
    "4. UTR-based prediction - We didn't train our approach for UTRs, ignore this table\n",
    "\n",
    "For each, we get different statistics, but let's focus on sensitivity and specificity. These two values are commonly\n",
    "used in test-training approaches and in machine learning like the one we have used here. For our exon level data, we\n",
    "should have a table similar to this (the actual numbers will be different):\n",
    "\n",
    "![](images/gene_1.png)\n",
    "\n",
    "Now what do the values in this table mean in plain English:\n",
    "\n",
    "* Sensitivity - How did we do in identifying real exons?\n",
    "    * Out of the total number of actual true exons (159), we found 110 total true events:\n",
    "        * 122/159 = 76.7%\n",
    "* Specificity - How many exons did we identify that are not real exons?\n",
    "    * Out of the total number of predictions (144), we found 34 additional events that are not actually exons:\n",
    "        * 1 - (29 / 151) = 80.8%\n",
    "\n",
    "Overall, this is a pretty good start for identifying exons. You may have noticed in one of the other tables that our\n",
    "ability to predict entire transcripts is not nearly as good. We will visually evalute the results at the end of this\n",
    "section to see why that may be.\n",
    "\n",
    "**Question:** What do you think a limitation of using just 1 chromosome to train our gene finder is?"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Identifying Genes\n",
    "\n",
    "Let's quickly review what we have done:\n",
    "\n",
    "1. Aligned RNA-seq data to our assembly\n",
    "2. Identified putative introns in these alignments\n",
    "3. Trained a model that tells Augustus what a gene in our assembly looks like.\n",
    "\n",
    "So, all that is left is to:\n",
    "\n",
    "4. Identify genes\n",
    "\n",
    "Now to actually run Augustus on our assembly:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "augustus --species=mygenome PB.masked.fasta > PB.contigs.gff"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "This command will likely take a minute or two depending on the computer you are using.\n",
    "\n",
    "We are also going to run Augustus using their default parameters, trained for _P. falciparum_, to compare against:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "augustus --species=pfalciparum PB.masked.fasta > PB.contigs.default.gff"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "Each of these commands will generate a single gff format file. The ENSEMBL website has more information on the\n",
    "format of [gff files](https://www.ensembl.org/info/website/upload/gff.html).\n",
    "\n",
    "**Questions:**\n",
    "\n",
    "1. Can you figure out how many genes each approach found?\n",
    "2. We identified protein coding genes, but can you think of any other types of annotations we could find with Augustus?\n",
    "\n",
    "> _hint: look back at the introduction to this practical._\n",
    "\n",
    "If these commands didn't work or took too long to run, we have placed backups at `annotation_backups/PB.contigs*.gff`"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Checking our Annotations in IGV\n",
    "\n",
    "Now that we have some genes, lets use the [Integrative Genomics Viewer](https://software.broadinstitute.org/software/igv/)\n",
    "to explore our results. You should have used IGV as part of several of the modules of this course, including RNA-seq\n",
    "and CHIP-seq. We will be using a similar approach here.\n",
    "\n",
    "Go ahead and open IGV:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "igv &"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "The `&` symbol tells IGV to open in another window on your virtual machine's desktop, but allows you to continue using\n",
    "your terminal. We are going to load your assembly as a genome in IGV.\n",
    "\n",
    "Now, load your assembly, RNA-seq alignments, and gene data into IGV:\n",
    "\n",
    "1. **Go to _\"Genomes -> Load Genome From File...\"_. Select \"PB.masked.fasta\" and click _\"Open\"_.**\n",
    "2. **Go to _\"File -> Load From File\"_. Select \"Pfalc.sorted.ssf.bam\" and click _\"Open\"_.**\n",
    "3. **Go to _\"File -> Load From File\"_. Select \"PB.contigs.gff\" and click _\"Open\"_.**\n",
    "4. **Go to _\"File -> Load From File\"_. Select \"PB.contigs.default.gff\" and click _\"Open\"_.**\n",
    "\n",
    "If you cannot find your files, make sure to navigate to the folder we are working in:\n",
    "\n",
    "Go to **course_data/ --> annotation/ --> data/**\n",
    "\n",
    "You should now have your assembly loaded into IGV, a set of RNA-seq reads labeled \"Pfalc.sorted.bam Coverage\" and\n",
    "\"Pfalc.sorted.bam\" and two gene annotations. Make sure you \"_right click_\" on the gene annotation and click _\"Squished\"_\n",
    "so that you can see exon-intron information. If you \"Zoom in\", your window should look similar to:\n",
    "\n",
    "![](images/IGV_1.png)\n",
    "\n",
    "Now, navigate to the top of the window and go to the coordinates:\n",
    "\n",
    "    tig00000001:77,000-80,500\n",
    "\n",
    "**Questions:**\n",
    "\n",
    "1. How many exons does this gene have?\n",
    "2. Do you think there are any issues in the gene structure?\n",
    "3. How do the predictions for your model versus the default model compare?\n",
    "\n",
    "Now go to the coordinates:\n",
    "\n",
    "    tig00000001:123,400-126,800\n",
    "\n",
    "**Question:** What are both predictions missing, and why do you think that is?\n",
    "\n",
    "> _hint: look back at the introduction to this practical._\n",
    "\n",
    "Finally, go to the coordinates:\n",
    "\n",
    "    tig00000001:165,000-171,500\n",
    "\n",
    "If you click on the gene structure (the one with exons and introns) of the gene in IGV, you should see a window like\n",
    "the following:\n",
    "\n",
    "![](images/IGV_2.png)\n",
    "\n",
    "The \"name\" of this gene in the window is g54, which simply means that it is the 54th gene that Augustus identified.\n",
    "Note that the name of your gene may be slightly different - please adjust the following commands accordingly.\n",
    "We can then go back to our terminal and find this gene in our .gff file using a simple command like grep:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "grep -A 30 \"start gene g54\" PB.contigs.gff"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "execution_count": null,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "* `-A` tells us how many lines \"after\" we match g54 we should report.\n",
    "\n",
    "Make sure to modify your command if the gene is not named \"g54\". Note that this command will not work for every gene\n",
    "that we identified exactly as written as they will have a different length (modifiable with the `-A` parameter) and a\n",
    "different name.\n",
    "\n",
    "**Questions:**\n",
    "\n",
    "1. How many exons does this gene have?\n",
    "2. How many introns?\n",
    "3. How can you extract the same information for another gene by modifying the above command? Report the command and result here.\n",
    "\n",
    "Feel free to explore the data in IGV before you continue to the final module,\n",
    "[Comparative Genomics](comparative_genomics.ipynb)."
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}